interpreter/prune.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/common/operators"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+)
+
+type astPruner struct {
+	expr  *expr.Expr
+	state EvalState
+}
+
+// TODO Consider having a separate walk of the AST that finds common
+// subexpressions. This can be called before or after constant folding to find
+// common subexpressions.
+
+// Prunes the given AST based on the given EvalState and generates a new AST.
+// Given AST is copied on write and a new AST is returned.
+// Couple of typical use cases this interface would be:
+//
+// A)
+// 1) Evaluate expr with some unknowns,
+// 2) If result is unknown:
+//   a) PruneAst
+//   b) Goto 1
+// Functional call results which are known would be effectively cached across
+// iterations.
+//
+// B)
+// 1) Compile the expression (maybe via a service and maybe after checking a
+//    compiled expression does not exists in local cache)
+// 2) Prepare the environment and the interpreter. Activation might be empty.
+// 3) Eval the expression. This might return unknown or error or a concrete
+//    value.
+// 4) PruneAst
+// 4) Maybe cache the expression
+// This is effectively constant folding the expression. How the environment is
+// prepared in step 2 is flexible. For example, If the caller caches the
+// compiled and constant folded expressions, but is not willing to constant
+// fold(and thus cache results of) some external calls, then they can prepare
+// the overloads accordingly.
+func PruneAst(expr *expr.Expr, state EvalState) *expr.Expr {
+	pruner := &astPruner{
+		expr:  expr,
+		state: state}
+	_, newExpr := pruner.prune(expr)
+	return newExpr
+}
+
+func (p *astPruner) createLiteral(node *expr.Expr, val *expr.Literal) *expr.Expr {
+	newExpr := *node
+	newExpr.ExprKind = &expr.Expr_LiteralExpr{LiteralExpr: val}
+	return &newExpr
+}
+
+func (p *astPruner) maybePruneAndOr(node *expr.Expr) (bool, *expr.Expr) {
```

Comment:
Typically the value is communicated first, and the state second, e.g.
`func (p *astPruner) maybePruneAndOr(...) (*expr.Expr, bool)`

Reply:
Done.